dattri/task.py

```diff
@@ -0,0 +1,238 @@
+
+        Args:
+            index (int): The index of the checkpoint to be loaded.
+        """
+        if self.current_checkpoint_idx is None or self.current_checkpoint_idx != index:
+            if isinstance(self.checkpoints[index], str):
+                self.model.load_state_dict(torch.load(self.checkpoints[index]))
+            else:
+                self.model.load_state_dict(self.checkpoints[index])
+            self.current_checkpoint_idx = index
+            self.named_parameters = {
+                k: p for k, p in self.model.named_parameters() if p.requires_grad
+            }
+
+    @staticmethod
+    def _genearte_param_layer_map(
+        named_parameters: Dict[str, torch.Tensor],
+    ) -> List[int]:
+        """This function generate the param_layer_map automatically.
+
+        Args:
+            named_parameters (Dict[str, torch.Tensor]):
+                The named parameters of the model.
+
+        Returns:
+            List[int]: The map from the parameter to the layer.
+        """
+        named_parameters_keys = named_parameters.keys()
+        param_layer_map = []
+
+        current_layer = None
+        current_index = -1
+        for key in named_parameters_keys:
+            layer_name = ".".join(key.split(".")[:-1])
+            if layer_name != current_layer:
+                current_index += 1
+                current_layer = layer_name
+            param_layer_map.append(current_index)
+
+        return param_layer_map
+
+    def get_grad_target_func(
+        self,
+        layer_name: Optional[Union[str, List[str]]] = None,
+    ) -> Callable:
+        """Return a function that computes the gradient of the target function.
+
+        TODO: support partial parameter gradient.
+
+        Args:
+            layer_name (Optional[Union[str, List[str]]]): This has not been supported.
+
+        Returns:
+            Callable: The function that computes the gradient of the target function.
+
+        Raises:
+            NotImplementedError: If layer_name is not None.
+        """
+        if layer_name is not None:
+            error_msg = "layer_name has not been implemented yet."
+            raise NotImplementedError(error_msg)
+        return self.grad_func
+
+    def get_target_func(
+        self,
+        layer_name: Optional[Union[str, List[str]]] = None,
+    ) -> Callable:
+        """Return a function that computes the gradient of the target function.
+
+        TODO: support partial parameter gradient.
+
+        Args:
+            layer_name (Optional[Union[str, List[str]]]): This has not been supported.
+
+        Returns:
+            Callable: The target function itself.
+
+        Raises:
+            NotImplementedError: If layer_name is not None.
+        """
+        if layer_name is not None:
+            error_msg = "layer_name has not been implemented yet."
+            raise NotImplementedError(error_msg)
+        return self.target_func
+
+    def get_param(
```

Comment:
No it's generally for all TDA algorithms. since TDA developers will not have touch to the model anymore, they rely on the api provided by the `task` to get the parameters.